### PR TITLE
[PDE-251] [Fix] ottersec advisory check issues

### PR DESCRIPTION
**Pull request by ungaro**
## What's new in this PR?

1. `accrueYield` iterates through potentially all of `YieldDistributionTokenStorage::deposits` to calculate the amount of yield a `user` is owed. there’s a hacky `gasleft` to break out of the loop, but it seems unstable and creates a dependency on gas costs. importantly, it also means that the yield might not be fully up to date after `accrueYield`, creating a strange edgecase. can you use a more traditional yield per token model to calculate the amount of yield owed?
`Using yield per token model now.`

2. do you expect `AssetTokenStorage::{whitelist, holders}` to ever contain a non-trivial amount of addresses?
`removed arrays and loops, added events so they can be indexed`

some other general questions:

1. what’s the intention of using the `from.call(abi.encodeWithSelector(…))` pattern? why not just cast `from` to `ISmartWallet` and do the function call directly?
     `The Solidity compiler adds a check that the target address has extcodesize > 0
        and otherwise reverts for high-level calls, so we have to use a low-level call here`

2. is there sample code using `_notifyRedeem` and `_notifyDeposit`? the example `USDT.sol` doesn’t use these functions, and they seem unused throughout the rest of the codebase

3. it seems like `whitelist` isn’t used anywhere in the code. would it make sense to use.
`removed whitelist array`

### CODE REVIEW

File: smart-wallets/src/interfaces/IYieldDistributionToken.sol [hunk 1]
```diff
@@ -5,15 +5,86 @@ import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
...
+     * @notice Emitted when yield is accrued to a user's balance
+     * @param user Address of the user who accrued the yield
+     * @param currencyTokenAmount Amount of CurrencyToken accrued as yield
+     */
+    event YieldAccrued(address indexed user, uint256 currencyTokenAmount);
+
+    /// @notice Indicates a failure because a yield deposit is made in the same block as the last one
+    error DepositSameBlock();
+
+    /// @notice Indicates a failure because the transfer of CurrencyToken failed
```
Comment (eyqs) [inline smart-wallets/src/interfaces/IYieldDistributionToken.sol]: @param user Address of the user whose transfer failed
@param currencyTokenAmount Amount of CurrencyToken to be transferred

File: smart-wallets/src/token/AssetToken.sol [hunk 2]
```diff
@@ -27,8 +27,6 @@ contract AssetToken is WalletUtils, YieldDistributionToken, IAssetToken {
     struct AssetTokenStorage {
         /// @dev Total value of all circulating AssetTokens
         uint256 totalValue;
-        /// @dev Whitelist of users that are allowed to hold AssetTokens
-        address[] whitelist;
```
Comment (eyqs) [inline smart-wallets/src/token/AssetToken.sol:31]: Sorry we should keep the whitelist so we can call `getWhitelist`

File: smart-wallets/src/token/AssetToken.sol [hunk 3]
```diff
@@ -145,11 +150,19 @@ contract AssetToken is WalletUtils, YieldDistributionToken, IAssetToken {
...
         }
 
-        if (!$.hasHeld[to]) {
-            $.holders.push(to);
-            $.hasHeld[to] = true;
+        if (to != address(0)) {
+            // Check if to address will become a new holder
+            if (balanceOf(to) == 0) {
+                // Currently has zero balance
+                emit HolderStatusChanged(to, true);
```
Comment (eyqs) [inline smart-wallets/src/token/AssetToken.sol:175]: I think we should keep `holders` and `hasHeld` and remove the `HolderStatusChanged`

File: smart-wallets/src/token/AssetToken.sol
Comment (eyqs) [inline smart-wallets/src/token/AssetToken.sol:1]: I think you can revert all the changes to this file

File: smart-wallets/src/token/YieldDistributionToken.sol
Comment (eyqs) [inline smart-wallets/src/token/YieldDistributionToken.sol:1]: Is this file severely out of date?

File: smart-wallets/src/token/AssetToken.sol (hunk 3 above)
Comment (ungaro) [reply smart-wallets/src/token/AssetToken.sol:175]: we can get those info from events? 
holders will contain every address that has ever held tokens, even if they no longer do and arrays consume unbounded storage, gas costs increase linearly with array size

File: smart-wallets/src/token/YieldDistributionToken.sol
Comment (ungaro) [reply smart-wallets/src/token/YieldDistributionToken.sol:1]: check 
https://github.com/plumenetwork/contracts/pull/91
https://github.com/plumenetwork/contracts/pull/54
https://github.com/plumenetwork/contracts/pull/67 (deleted, keeping here for comments)
https://github.com/plumenetwork/contracts/pull/129 (new 67)

**Comment by eyqs** [comment]
Discussed offline - better to make the smart contract smaller and rely on offchain event indexing.

File: smart-wallets/src/token/AssetToken.sol (hunk 2 above)
Comment (ungaro) [reply smart-wallets/src/token/AssetToken.sol:31]: we can handle this offchain.

**Review by eyqs — APPROVED** [review]
LGTM

**Pull request merged**